wordpress-mobile/WordPress-iOS
# Fix cached notifications appear when logout

Fixes #19075

| -- | Before | After |
| -- | -- | -- |
| iPhone | ![Simulator Screen Shot - iPhone 14 - 2022-09-27 at 21 13 09](https://user-images.githubusercontent.com/10288494/192629934-3f1ac4e9-8a4f-4a33-9297-c1db1159066a.png) | ![Simulator Screen Shot - iPhone 14 - 2022-09-28 at 04 25 57](https://user-images.githubusercontent.com/10288494/192630004-4af591e2-d742-492c-9abf-91d535bd916b.png) |
| iPad | ![Simulator Screen Shot - iPad (9th generation) - 2022-09-28 at 03 12 23](https://user-images.githubusercontent.com/10288494/193042545-16efc1cd-aa65-4cbe-8d4f-1b4f34a39ba6.png) | ![Simulator Screen Shot - iPad (9th generation) - 2022-09-28 at 03 49 36](https://user-images.githubusercontent.com/10288494/192630147-a1578d79-3427-4daa-904b-c400cfa81b89.png) |

## How To Test:
1. Lanch the app and log in with a WordPress.com account
2. Tap `My Site` tab, open the site-switch page 
3. Tap `+` and tap `Add self-hosted site`
4. Typing address and userName/password and login
5. Tap the `Notifications` tab and observe existing notifications
6. Tap `My Site` tab
7. Tap `user` icon button at top-right navigation bar, and tap `Log Out` to logout
8. Tap the Notifications tab and observe only the jetpack view without notification.

## Root Cause:
The notification cache didn't delete if the  `NotificationsViewController` is not [presented](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L689).

## Fix:
0d8cb1a7dc95fd6ba82f5c90459e629034843b5e  :
- delete the notification cache and other cache params no matter view is presented or not
- set `needsReloadResults = true` if the view is not presented. (it will lazy reload at the [viewWillAppear](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L196))
- use `reloadResultsController` instead of `tableView.reloadData` because the [sections](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L1170) also need reload.

d28c64aadaefa0e3eea3e8cfeed79754790a603c  :
- fix the `JetpackLoginViewController` to align the center of tableView.

| before | after |
| -- | -- |
| ![Simulator Screen Shot - iPhone 14 - 2022-09-28 at 01 58 28](https://user-images.githubusercontent.com/10288494/192814529-46ebed8a-8439-4ce8-ba56-8013cd3e7369.png) | ![Simulator Screen Shot - iPhone 14 - 2022-09-28 at 02 33 37](https://user-images.githubusercontent.com/10288494/192814614-1e56decb-8b01-4e99-a228-0bbd56bd8559.png) |

- The `JetpackLoginViewController` will have labels overlapping in large `Dynamic Type`. It should open another issue/PR to handle it.

dd7bff1cdd1a0c5ba2345a7cdaa0f0970a682e82  :
- On iPad, the `detailViewController` should be clean at [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L1320).  But it doesn't clean correctly because the `splitViewController.wpPrimaryColumnWidth` is set to `.full` at [here](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.swift#L1561). 
- We should not use the `.full` mode because it does nothing change by the [reason](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/System/WPSplitViewController.swift#L74). You will stay `default` view style even when you select `.full`.
- So this commit adds a comment and remove the `.full` mode temporarily.
- the [BlogListViewController](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/ViewRelated/Blog/Blog%20List/BlogListViewController.m#L348) also has similar `.full` logic, it should open a new issue to check if other views work fine on `.full` mode.

| before | after |
| -- | -- |
| ![Simulator Screen Shot - iPad (9th generation) - 2022-09-28 at 03 12 23](https://user-images.githubusercontent.com/10288494/193042545-16efc1cd-aa65-4cbe-8d4f-1b4f34a39ba6.png) | ![Simulator Screen Shot - iPad (9th generation) - 2022-09-28 at 02 45 51](https://user-images.githubusercontent.com/10288494/192815572-244516a2-1baa-497c-82e5-0556ae3aaffa.png) |

4d997564752e1aa1351add708b3af0e0dafb2c1f 
799c547c31968bfc73488a3c6f6b35d7f7ef9246 :
- Tweak the logic of `shouldDisplayJetpackPrompt`, it should check `blog != nil` before presenting it. Otherwise, the jetpackView will not show and will get an empty page.

a92f36d5f4461e91a7ff5bb869f6182c6d7a9e7e 
776edf09d07afcb7e2b907db406a7f2ccf23ce99 :
- Modified the `NotificationsViewController` interface and made it testable.
- Added unit tests for testing the parameter reset behavior when the account changed. 

## Regression Notes
1. Potential unintended areas of impact
`Notifications` tab on iPad/iPhone

3. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing on the `Notification` tab when login/logout WP.com account. (both iPad/iPhone)

4. What automated tests I added (or what prevented me from doing so)
614f37450c837af60625c5089173d4612f0b9c3a Added unit tests for testing the parameter reset behavior when the account changed. 

PR submission checklist:
